### ISSUE TITLE
PowerShell ValueFromPipeline parameter breaks ParameterSet selection

### ISSUE THREAD

**Issue by ili101** [PowerShell/PowerShell#8516]
# Steps to reproduce
```PowerShell
function Export-Excel {
    [CmdletBinding(DefaultParameterSetName = 'Default')]
    Param(
        [Parameter(ParameterSetName = "Default")]
        [Parameter(ParameterSetName = "AutoFilter")]
        [String]$Path,

        [Parameter(Mandatory = $true, ParameterSetName = "PackageDefault")]
        [Parameter(Mandatory = $true, ParameterSetName = "PackageAutoFilter")]
        $ExcelPackage,

        [Parameter(ValueFromPipeline = $true)]
        $TargetData,

        [Parameter(Mandatory = $true, ParameterSetName = 'AutoFilter')]
        [Parameter(Mandatory = $true, ParameterSetName = 'PackageAutoFilter')]
        $AutoFilter,

        [Parameter(ParameterSetName = 'Default')]#, ValueFromPipelineByPropertyName)]
        [Parameter(ParameterSetName = 'PackageDefault')]#, ValueFromPipelineByPropertyName)]
        [String]$TableName,
        [Parameter(ParameterSetName = 'Default')]
        [Parameter(ParameterSetName = 'PackageDefault')]
        [Switch]$Table,
        [Parameter(ParameterSetName = 'Default')]
        [Parameter(ParameterSetName = 'PackageDefault')]
        $TableStyle = 'Medium6'
    )
    "ParameterSet: " + $PSCmdlet.ParameterSetName
}
# Works
Export-Excel -ExcelPackage 5 -TargetData 5
Export-Excel -ExcelPackage 5
# Prompt for "AutoFilter:"
5 | Export-Excel -ExcelPackage 5
```

# Expected behavior
```none
ParameterSet: PackageDefault
ParameterSet: PackageDefault
ParameterSet: PackageDefault
```

# Actual behavior
```none
ParameterSet: PackageDefault
ParameterSet: PackageDefault

cmdlet Export-Excel at command pipeline position 1
Supply values for the following parameters:
AutoFilter:
```

When not passing $ExcelPackage or passing it as a parameter ParameterSet PackageDefault is selected, but when providing it from Pipeline it asks for $AutoFilter from the wrong ParameterSet.

Tested on 5.1 and 6.2.0-preview.3

I also tried to workaround this by providing multiple DefaultParameterSetName hoping it will evaluate them in order but looks like it's not a thing, Feature request?
```PowerShell
[CmdletBinding(DefaultParameterSetName = 'Default,', 'PackageDefault')]
```
`Cannot find an overload for ".ctor" and the argument count: "1".`

**Comment by ili101** [comment]
Apparently adding `ValueFromPipelineByPropertyName` to `$AutoFilter` workaround this problem (I have no idea why though, It doesn't make sense)
``` PowerShell
[Parameter(Mandatory = $true, ParameterSetName = 'AutoFilter', ValueFromPipelineByPropertyName)]
[Parameter(Mandatory = $true, ParameterSetName = 'PackageAutoFilter', ValueFromPipelineByPropertyName)]
$AutoFilter,
```

**Comment by mklement0** [comment]
Good find, @ili101. To make this issue easier to discover, please fix the typo in the title: _brakes_ -> _breaks_.

**Comment by ili101** [comment]
Any update on this problem?
Also the `ValueFromPipelineByPropertyName` workaround I described above is problematic as it preventing the `$PSCmdlet.ParameterSetName` from being set in the `begin{}` part:

``` PowerShell
function Export-Excel
{
    [CmdletBinding(DefaultParameterSetName = 'Default')]
    Param(
        [Parameter(ParameterSetName = "Default")]
        [Parameter(ParameterSetName = "AutoFilter")]
        [String]$Path,

        [Parameter(Mandatory = $true, ParameterSetName = "PackageDefault")]
        [Parameter(Mandatory = $true, ParameterSetName = "PackageAutoFilter")]
        $ExcelPackage,

        [Parameter(ValueFromPipeline = $true)]
        $TargetData,

        [Parameter(Mandatory = $true, ParameterSetName = 'AutoFilter', ValueFromPipelineByPropertyName)]
        [Parameter(Mandatory = $true, ParameterSetName = 'PackageAutoFilter', ValueFromPipelineByPropertyName)]
        $AutoFilter,

        [Parameter(ParameterSetName = 'Default')]#, ValueFromPipelineByPropertyName)]
        [Parameter(ParameterSetName = 'PackageDefault')]#, ValueFromPipelineByPropertyName)]
        [String]$TableName,
        [Parameter(ParameterSetName = 'Default')]
        [Parameter(ParameterSetName = 'PackageDefault')]
        [Switch]$Table,
        [Parameter(ParameterSetName = 'Default')]
        [Parameter(ParameterSetName = 'PackageDefault')]
        $TableStyle = 'Medium6'
    )
    begin {"ParameterSet begin: " + $PSCmdlet.ParameterSetName}
    process {"ParameterSet process: " + $PSCmdlet.ParameterSetName}
    end {"ParameterSet end: " + $PSCmdlet.ParameterSetName}
}
# Works
Export-Excel -ExcelPackage 5 -TargetData 5
Export-Excel -ExcelPackage 5
# With ValueFromPipelineByPropertyName workaround no ParameterSetName in begin:
5 | Export-Excel -ExcelPackage 5
```
Output:
```
ParameterSet begin: PackageDefault
ParameterSet process: PackageDefault
ParameterSet end: PackageDefault
ParameterSet begin: PackageDefault
ParameterSet process: PackageDefault
ParameterSet end: PackageDefault
ParameterSet begin:
ParameterSet process: PackageDefault
ParameterSet end: PackageDefault
```

**Comment by mklement0** [comment]
I've created a new issue focused on the core symptom of your code, which I think is a bug: #11235

The workaround with `ValueFromPipelineByPropertyName` can have unexpected side effects (pipeline input may suddenly start binding to your `$AutoFilter` parameter, which is undesired).

> as it is preventing the `$PSCmdlet.ParameterSetName` from being set in the `begin{}` part:

While I have no explanation for why `$PSCmdlet.ParameterSetName` is the _empty string_ in the `begin` block with your workaround (smells like another bug), it is worth noting that in functions that have  pipeline-binding parameters, `$PSCmdlet.ParameterSetName` generally only has a meaningful value in the `process` block, because it is possible for _each input object_ to select a _different_ parameter set and the value in `begin` seems to reflect the set that _would have been chosen had there been no pipeline input_.

In short: To be safe, check `$PSCmdlet.ParameterSetName`  only in the `process` block.

**Comment by mklement0** [comment]
Additionally, #11237 proposes fundamental - but breaking - improvements to how parameter sets are handled.

If you agree that #11235 and #11237 cover all aspects of this issue, I encourage you to close this issue.
